File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/taskdetail/TaskDetailPresenter.java
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.taskdetail;

import android.support.annotation.Nullable;

import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestackdemoexamplefragments.data.repository.TaskRepository;
import com.zhuinden.simplestackdemoexamplefragments.presentation.objects.Task;
import com.zhuinden.simplestackdemoexamplefragments.presentation.paths.addoredittask.AddOrEditTaskKey;
import com.zhuinden.simplestackdemoexamplefragments.util.BasePresenter;
import com.zhuinden.simplestackdemoexamplefragments.util.Strings;
import com.zhuinden.statebundle.StateBundle;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;

/**
 * Created by Owner on 2017. 01. 27..
 */

public class TaskDetailPresenter
        extends BasePresenter<TaskDetailFragment, TaskDetailPresenter> {
    private final TaskRepository taskRepository;
    private final Backstack backstack;

    @Inject
    public TaskDetailPresenter(TaskRepository taskRepository, Backstack backstack) {
        this.taskRepository = taskRepository;
        this.backstack = backstack;
    }

    TaskDetailKey taskDetailKey;

    String taskId;

    Task task;

    @Override
    protected void onAttach(TaskDetailFragment coordinator) {
        taskDetailKey = coordinator.getKey();
        this.taskId = taskDetailKey.taskId();
        taskRepository.findTask(taskId).observeOn(AndroidSchedulers.mainThread()).subscribe(taskOptional -> {
            if(taskOptional.isPresent()) {
                task = taskOptional.get();
                coordinator.showTask(task);
            } else {
                task = null;
                coordinator.showMissingTask();
            }
        });
    }

    @Override
    protected void onDetach(TaskDetailFragment coordinator) {

    }

    public void editTask() {
        if(Strings.isNullOrEmpty(taskId)) {
            getFragment().showMissingTask();
            return;
        }
        backstack.goTo(AddOrEditTaskKey.createWithTaskId(getFragment().getKey(), taskId));
    }

    public void completeTask(Task task) {
        taskRepository.setTaskCompleted(task);
    }

    public void activateTask(Task task) {
        taskRepository.setTaskActive(task);
    }

    public void deleteTask() {
        if(task != null) {
            taskRepository.deleteTask(task);
            backstack.goBack();
        }
    }

    @Override
    public StateBundle toBundle() {
        return new StateBundle();
    }

    @Override
    public void fromBundle(@Nullable StateBundle bundle) {
    }
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/taskdetail/TaskDetailPresenter.kt
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.taskdetail

import android.annotation.SuppressLint
import com.zhuinden.simplestack.Backstack
import com.zhuinden.simplestackdemoexamplefragments.data.repository.TaskRepository
import com.zhuinden.simplestackdemoexamplefragments.presentation.objects.Task
import com.zhuinden.simplestackdemoexamplefragments.presentation.paths.addoredittask.AddOrEditTaskKey
import com.zhuinden.simplestackdemoexamplefragments.util.BasePresenter
import io.reactivex.android.schedulers.AndroidSchedulers
import javax.inject.Inject

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

class TaskDetailPresenter @Inject constructor(
    private val taskRepository: TaskRepository,
    private val backstack: Backstack
) : BasePresenter<TaskDetailFragment>(), TaskDetailFragment.Presenter {
    override fun onTaskChecked(task: Task, checked: Boolean) {
        if (checked) {
            completeTask(task)
        } else {
            activateTask(task)
        }
    }

    override fun onTaskEditButtonClicked() {
        editTask()
    }

    override fun onTaskDeleteButtonClicked() {
        deleteTask()
    }

    lateinit var taskDetailKey: TaskDetailKey

    lateinit var taskId: String

    var task: Task? = null

    @SuppressLint("CheckResult")
    override fun onAttach(view: TaskDetailFragment) {
        this.taskDetailKey = view.getKey()
        this.taskId = taskDetailKey.taskId

        taskRepository.findTask(taskId)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { taskOptional ->
                val task = taskOptional.takeIf { it.isPresent }?.get()
                this.task = task
                if(task != null) {
                    view.showTask(task)
                } else {
                    view.showMissingTask()
                }
            }
    }

    override fun onDetach(view: TaskDetailFragment) {

    }

    private fun editTask() {
        if (task == null) {
            view?.showMissingTask()
            return
        }
        backstack.goTo(AddOrEditTaskKey.EditTaskKey(view!!.getKey(), taskId))
    }

    private fun completeTask(task: Task) {
        taskRepository.setTaskCompleted(task)
    }

    private fun activateTask(task: Task) {
        taskRepository.setTaskActive(task)
    }

    private fun deleteTask() {
        val task = task
        if (task != null) {
            taskRepository.deleteTask(task)
            backstack.goBack()
        }
    }
}
